protocol/auditlog.go

```diff
@@ -0,0 +1,196 @@
+// This module implements a CONIKS audit log that a CONIKS auditor
+// maintains.
+// An audit log is a mirror of many CONIKS key directories' STR history,
+// allowing CONIKS clients to audit the CONIKS directories.
+
+package protocol
+
+import (
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	m "github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type directoryHistory struct {
+	signKey   sign.PublicKey
+	snapshots map[uint64]*m.SignedTreeRoot
+	latestSTR *m.SignedTreeRoot
+}
+
+// A ConiksAuditLog maintains the histories
+// of all CONIKS directories known to a CONIKS auditor.
+// Each history includes the directory's public signing key
+// enabling the auditor to verify the corresponding signed
+// tree roots.
+type ConiksAuditLog struct {
+	histories map[string]*directoryHistory
+}
+
+func newDirectoryHistory(signKey sign.PublicKey, str *m.SignedTreeRoot) *directoryHistory {
+	h := new(directoryHistory)
+	h.signKey = signKey
+	h.snapshots = make(map[uint64]*m.SignedTreeRoot)
+	h.latestSTR = str
+	return h
+}
+
+// NewAuditLog constructs a new ConiksAuditLog. It creates an empty
+// log; the auditor will add an entry for each CONIKS directory
+// the first time it observes an STR for that directory.
+func NewAuditLog() *ConiksAuditLog {
+	l := new(ConiksAuditLog)
+	l.histories = make(map[string]*directoryHistory)
+	return l
+}
+
+// IsKnownDirectory checks to see if an entry for the directory
+// address addr exists in the audit log l. IsKnownDirectory() does not
+//  validate the entries themselves. It returns true if an entry exists,
+// and false otherwise.
+func (l *ConiksAuditLog) IsKnownDirectory(addr string) bool {
+	h := l.histories[addr]
+	if h != nil {
+		return true
+	}
+	return false
+}
+
+// FIXME: pass Request message as param
```

Comment:
Ah, you're right, this should be `Response`. I also like the idea of having a generic `HandleDirectorySTRs` function since there is a lot of potentially shared STR verification code between the auditor and the client.  But `Insert`, `Update` and `InsertRange` are specific to the auditor log, though, so I think it makes sense to keep those separate from `consistencychecks.go`.